tools/blocksconvert/planprocessor/service.go

```diff
@@ -0,0 +1,405 @@
+package planprocessor
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	"github.com/cortexproject/cortex/pkg/util/grpcclient"
+	"github.com/cortexproject/cortex/pkg/util/services"
+	"github.com/cortexproject/cortex/tools/blocksconvert"
+)
+
+type PlanProcessor interface {
+	// Returns "id" that is appended to "finished" status file.
+	ProcessPlanEntries(ctx context.Context, entries chan blocksconvert.PlanEntry) (string, error)
+}
+
+type Config struct {
+	// Exported config options.
+	Name              string
+	HeartbeatPeriod   time.Duration
+	SchedulerEndpoint string
+	NextPlanInterval  time.Duration
+	GrpcConfig        grpcclient.ConfigWithTLS
+
+	// Additional settings required for plan processor service.
+
+	PlansDirectory string                                                                            // Where to store plan files.
+	Bucket         objstore.Bucket                                                                   // Bucket client used for downloading plan files.
+	Cleanup        func(logger log.Logger) error                                                     // Cleanup function called on startup and after each build. Can be nil.
+	Factory        func(planLog log.Logger, userID string, dayStart, dayEnd time.Time) PlanProcessor // Factory for creating PlanProcessor. Called for each new plan.
```

Comment:
Design wise, I would have expected these to be arguments to `NewService()`. Thoughts?